Documentation/Contributing.md

```diff
@@ -0,0 +1,283 @@
+# Contributing to Swift Package Manager
+There are several types of contributions one can make. Bug fixes, documentation and enhancements that do not materially change the user facing semantics of Swift Package Manager should be submitted directly as PR.
+
+Larger changes that do materially change the semantics of Swift Package Manager (e.g. changes to the manifest format or behavior) are required to go through [Swift Evolution Process](https://github.com/apple/swift-evolution/blob/master/process.md).
+
+To see how previous evolution decisions for SwiftPM have been made and have some direction for the development of future features please check out the [Community Proposals](https://forums.swift.org/tag/packagemanager).
+
+For more information about making contributions to the Swift project in general see [Swift Contribution Guide](https://swift.org/contributing/).
+
+## Reporting issues
+* [SwiftPM JIRA Bug Tracker](https://bugs.swift.org/browse/SR-13640?jql=component%20%3D%20%22Package%20Manager%22)
+* [Guide for filing quality bug reports](https://github.com/apple/swift-package-manager/blob/main/Documentation/Resources.md#reporting-a-good-swiftpm-bug)
+
+Report a bug guide: https://github.com/apple/swift-package-manager/blob/main/Documentation/Resources.md#reporting-a-good-swiftpm-bug.  
+JIRA Bug Tracker (a place where you can open bugs, enhancements to start to contribute): [https://bugs.swift.org/browse/SR-13640?jql=component%20%3D%20%22Package%20Manager%22](https://bugs.swift.org/browse/SR-13640?jql=component%20%3D%20%22Package%20Manager%22).
+
+## Development environment
```

Comment:
here is an improved version we can use for this while section:

```
## Development environment

First, clone a copy of SwiftPM code from https://github.com/apple/swift-package-manager. 

If you are preparing to make a contribution you should fork the repository first and clone the fork which will make opening Pull Requests easier. See "Creating Pull Requests" section below.

SwiftPM is typically built with a pre-existing version of SwiftPM present on the system, but there are multiple ways to setup your development environment:

### Using Xcode (Easiest)

1. Install Xcode from [https://developer.apple.com/xcode](https://developer.apple.com/xcode) (including betas!).
2. Verify the expected version of Xcode was installed.
3. Open SwiftPM's `Package.swift` manifest with Xcode, and use Xcode to edit the code, build, and run the tests.

### Using the Command Line

If you are using macOS and have Xcode installed, you can use the command line even without downloading and installing a toolchain, otherwise, you first need to doownload and install one.

#### Installing the toolchain

1. Download a toolchain from https://swift.org/download/
2. Install it and verify the expected version of the toolchain was installed:

  **macOS**
  ```bash
  $> export TOOLCHAINS=swift
  ```

  Verify that we're able to find the swift compiler from the installed toolchain.
  ```bash
  $> xcrun --find swift
  /Library/Developer/Toolchains/swift-latest.xctoolchain/usr/bin/swift
  $> swift package --version
  Swift Package Manager - Swift 5.3.0
  $> swift --version
  Apple Swift version 5.3
  ```

  **Linux**
  ```bash
  $> export PATH=/path/to/swift-toolchain/usr/bin:"${PATH}"
  ```
  Verify that we're able to find the swift compiler from the installed toolchain.
  ```bash
  $> which swift
  /path/to/swift-toolchain/usr/bin/swift
  $> swift package --version
  Swift Package Manager - Swift 5.3.0
  $> swift --version
  Apple Swift version 5.3
  ```
  
Note:  Alternatively use tools like [swiftenv](https://github.com/kylef/swiftenv) that help manage toolchains versions.

#### Building

```bash
$> swift build
```

A successful build will create a `.build/` directory with the following approximate structure:
```bash
artifacts/
checkouts/
debug/
repositories
x86_64-apple-macosx
```

Binary artifacts are located in `x86_64-apple-macosx/` when building on macOS, 
or the equivalent on other architectures and operating systems.

These binaries can be used to test the code modification. For example, to test the `swift package init` and `swift build` commands from the new SwiftPM artifacts in `.build/`:

```bash
$> cd /tmp && mkdir hello && cd hello
$> /path/to/swiftpm/.build/x86_64-apple-macosx/debug/swift-package init
$> /path/to/swiftpm/.build/x86_64-apple-macosx/debug/swift-build
```

#### Testing

```bash
$> swift test
```

to run a single test:

```bash
$> swift test --filter PackageGraphTests.DependencyResolverTests/testBasics
```

Or another example, to run tests for the test targets BuildTests and WorkspaceTests, but skip some test cases:

```bash
$> swift test --filter BuildTests --skip BuildPlanTests --filter WorkspaceTests --skip InitTests
```

To run the performance tests, enable them with an ENV variable:

```bash
$> export TSC_ENABLE_PERF_TESTS=1
$> swift test -c release --filter PerformanceTests
```

### Using the bootstrap script

The bootstrap script is designed for building SwiftPM on systems that do not have Xcode or a toolchain installed.
It is used on bare systems to bootstrap the Swift toolchain (including SwiftPM), and as such not typically used outside the Swift team.

the bootstrap script requires having [CMake](https://cmake.org/) and [Ninja](https://ninja-build.org/) installed. 
Please refer to the [Swift project repo](https://github.com/apple/swift/blob/master/README.md#macos) for installation instructions.

1. Clone [llbuild](https://github.com/apple/swift-llbuild) beside the SwiftPM directory.

  ```bash
  $> git clone https://github.com/apple/swift-llbuild llbuild
  ```

  Note: Make sure the directory for llbuild is called "llbuild" and not
      "swift-llbuild".

2. Clone [Yams](https://github.com/jpsim/yams) beside the SwiftPM directory.

  ```bash
  $> git clone https://github.com/jpsim/yams
  ```

3. Clone [swift-driver](https://github.com/apple/swift-driver) beside the SwiftPM directory.

  ```bash
  $> git clone https://github.com/apple/swift-driver
```

4. Clone [swift-argument-parser](https://github.com/apple/swift-argument-parser) beside the SwiftPM directory and check out tag with the [latest version](https://github.com/apple/swift-argument-parser/tags).

  For example, if the latest tag is 0.3.1:
  ```bash
  $> git clone https://github.com/apple/swift-argument-parser --branch 0.3.1
  ```

#### Building

```bash
$> Utilities/bootstrap build
```

See "Using the Command Line / Building" section above for more information on how to test the new artifacts.

#### Testing

```bash
$> Utilities/bootstrap test
```

## Testing locally

Before submitting code modification as Pull Requests, test locally across the supported platforms and build variants.

1. If using Xcode, run all the unit tests and verify they pass.

2. If using the Command Line, run all the unit tests and verify they pass.

  ```bash
  $> swift test
  ```

3. Optionally: Test with the bootstrap script as well.

  ```bash
  $> Utilities/bootstrap test
  ```
  
When developing on macOS and need to test on Linux, install [Docker](https://www.docker.com/products/docker-desktop) and use the following commands:

```bash
$> Utilities/Docker/docker-utils build # will build an image with the latest Swift snapshot
$> Utilities/Docker/docker-utils bootstrap # will bootstrap SwiftPM on the Linux container
$> Utilities/Docker/docker-utils run bash # to run an interactive Bash shell in the container
$> Utilities/Docker/docker-utils swift-build # to run swift-build in the container
$> Utilities/Docker/docker-utils swift-test # to run swift-test in the container
$> Utilities/Docker/docker-utils swift-run # to run swift-run in the container
```    


```